src/clj/puppetlabs/services/master/master_core.clj

```diff
@@ -710,6 +712,48 @@
          :headers {"Content-Type" "application/octet-stream"}
          :body (get-code-content environment code-id file-path)}))))
 
+(defn decode-catalog-post-body
+  [body]
+  (let [parameters
+        (try+
+          (json/decode body true)
+          (catch JsonParseException e
+            (throw+ {:kind :bad-request
+                     :msg (format "Error parsing JSON: %s" e)})))]
+    (schema/validate
+     {:certname schema/Str
+      :persistence {:facts schema/Str, :catalogs schema/Str, :reports schema/Str}
+      (schema/optional-key :trusted_facts) {:values {schema/Any schema/Any}}
+      (schema/optional-key :facts) {:values {schema/Any schema/Any}}
+      (schema/optional-key :job-id) schema/Int
+      (schema/optional-key :environment) schema/Str
+      (schema/optional-key :classes) [schema/Any]
+      (schema/optional-key :parameters) {schema/Any schema/Any}}
+     parameters)
+    parameters))
+
+(schema/defn ^:always-validate
+  v4-catalog-fn :- IFn
+  [jruby-service :- (schema/protocol jruby-protocol/JRubyPuppetService)]
+  (fn [request]
+    (let [{:keys [certname persistence trusted_facts facts environment classes parameters]}
+          (decode-catalog-post-body (slurp (:body request)))]
+      {:status 200
+       :headers {"Content-Type" "application/json"}
+       :body (json/encode
+              (jruby-protocol/compile-catalog jruby-service
+                                              (:jruby-instance request)
+                                              {:certname certname
```

Comment:
This map should probably contain all the keys listed in the schema in the `decode-catalog-post-body` method?

Reply:
Agreed, I think they all end up being needed in the Ruby code.

Reply:
ok tony wrote that part